File: chriswynne/spec/canvas_spec.rb
require 'canvas.rb'

describe Canvas do
  let(:canvas) { described_class }

  describe '.new' do
    it 'creates a canvas with the correct height' do
      width = 1
      height = rand(1..10)
      new_canvas = canvas.new(width, height)
      expect(new_canvas.height).to eq(height)
    end

    it 'creates a canvas with the correct width' do
      width = rand(1..10)
      height = 1
      new_canvas = canvas.new(width, height)
      expect(new_canvas.width).to eq(width)
    end
  end

  describe '#draw' do
    it 'can return the correct width canvas' do
      new_canvas = canvas.new(4, 1)
      expect(new_canvas.draw).to eq(
        "------\n" \
        "|    |\n" \
        "------\n"
      )
    end

    it 'can return the correct height canvas' do
      new_canvas = canvas.new(1, 4)
      expect(new_canvas.draw).to eq(
        "---\n" \
        "| |\n" \
        "| |\n" \
        "| |\n" \
        "| |\n" \
        "---\n"
      )
    end
  end

  describe '#add_line' do

    it 'can add a horizontal line to the canvas' do
      new_canvas = canvas.new(4, 4)
      new_canvas.add_line(1, 3, 3, 3)
      expect(new_canvas.draw).to eq(
      "------\n" +
      "|    |\n" +
      "|    |\n" +
      "|xxx |\n" +
      "|    |\n" +
      "------\n"
      )
    end

    it 'can add a vertical line to the canvas' do
      new_canvas = canvas.new(4, 4)
      new_canvas.add_line(2, 1, 2, 3)
      expect(new_canvas.draw).to eq(
      "------\n" +
      "| x  |\n" +
      "| x  |\n" +
      "| x  |\n" +
      "|    |\n" +
      "------\n"
      )
    end

  end
end

File: andydowell/spec/canvasSpec.js
describe('canvas', function() {
  var Canvas = require('../js/canvas.js');
  var canvas; 

  var createNewCanvas = function(width, height) {
    var tempCanvas = new Canvas(width, height);
    tempCanvas.generateCanvas();
    return tempCanvas;
  }
  describe('on initialization', function() {
    it('should know it\'s dimensions', function() {
      var canvas = new Canvas(20, 4);
      expect(canvas.height).toEqual(4);
      expect(canvas.width).toEqual(20);
    })
  })

  describe('generateCanvas', function() {
    it('should generate a blank canvas', function() {
      expect(createNewCanvas(5,5).content[1][1]).toEqual(' ');
    })
  })

  describe('drawPoint', function() {
    it('should be able to draw a point', function() {
      canvas = createNewCanvas(5,5);
      canvas.drawPoint(2,3);
      expect(canvas.content[3][2]).toEqual('x');
    })
    it('shouldnt be able to draw a point outside of the canvas', function() {
      expect(function(){createNewCanvas(5,5).drawPoint(15,100)}).toThrow('Canvas is 5x5; Unable to draw point x:15 y:100 outside of canvas');
    })
  })

  describe('drawLine', function() {
    it('should be able to draw a horizontal line between two points', function() {
      canvas = createNewCanvas(5,5);
      canvas.drawLine(0,0,4,0);
      expect(canvas.content[0][0]).toEqual('x');
      expect(canvas.content[0][1]).toEqual('x');
      expect(canvas.content[0][2]).toEqual('x');
      expect(canvas.content[0][3]).toEqual('x');
      expect(canvas.content[1].indexOf('x')).toEqual(-1);
    })
    it('should be able to draw a vertical line between two points', function() {
      canvas = createNewCanvas(5,5);
      canvas.drawLine(0,0,0,2);
      expect(canvas.content[0][0]).toEqual('x');
      expect(canvas.content[1][0]).toEqual('x');
      expect(canvas.content[2][0]).toEqual('x');
      expect(canvas.content[3][0]).toEqual(' ');
    })
  })

  describe('renderCanvas', function() {
    it('should correctly render a blank square 5x5 canvas', function() {
      canvas = createNewCanvas(5,5);
      expect(canvas.renderCanvas()).toEqual(
					  '-------\n'+
					  '|     |\n'+
					  '|     |\n'+
					  '|     |\n'+
					  '|     |\n'+
					  '|     |\n'+
					  '-------\n'); 
    })
  })
})
